README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-users.svg)](https://packagist.org/packages/wdmg/yii2-users)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-users.svg)](https://packagist.org/packages/wdmg/yii2-users)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-users.svg)](https://github.com/wdmg/yii2-users/blob/master/LICENSE)

<img src="./docs/images/yii2-users.png" width="100%" alt="Yii2 Users Module" />

# Yii2 Users Module
Users management module

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-users"`

After configure db connection, run the following command in the console:

`$ php yii users/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations
  3) Batch insert demo data<sup>*</sup>

\* - The demo database contains 6 demo user`s with:

| ID   | Username  | Password   | Email               | Status      |
| ---- | --------- | ---------- | ------------------- | ----------- |
| 100  | admin     | admin      | admin@example.com   | `active`    |
| 101  | demo      | demo       | demo@example.com    | `inactive`  |
| 102  | alice     | alice      | alice@example.com   | `inactive`  |
| 103  | bob       | bob        | bob@example.com     | `inactive`  |
| 104  | johndoe   | johndoe    | johndoe@example.com | `inactive`  |
| 105  | janedoe   | janedoe    | janedoe@example.com | `inactive`  |

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-users/migrations`

# Configure

To add a module to the project, add the following data in your configuration file:

    
    'components' => [
        'user' => [
            'identityClass' => 'wdmg\users\models\Users',
        ],
        ...
    ],
    'modules' => [
        'users' => [
            'class' => 'wdmg\users\Module',
            'routePrefix' => 'admin',
            'rememberDuration' => (3600 * 24 * 30),
            'signupConfirmation' => [
                'needConfirmation' => false,
                'checkTokenRoute' => 'site/signup-confirm',
                'supportEmail' => 'noreply@example.com',
                'emailViewPath' => [
                    'html' => '@vendor/wdmg/yii2-users/mail/signupConfirmation-html',
                    'text' => '@vendor/wdmg/yii2-users/mail/signupConfirmation-text',
                ],
            ],
            'passwordReset' => [
                'resetTokenExpire' => 3600,
                'checkTokenRoute' => 'site/reset-password',
                'supportEmail' => 'noreply@example.com',
                'emailViewPath' => [
                    'html' => '@vendor/wdmg/yii2-users/mail/passwordReset-html',
                    'text' => '@vendor/wdmg/yii2-users/mail/passwordReset-text',
                ],
            ],
            'multiSignIn' => true,
            'sessionTimeout' => 0, // where `0` is unlimited
        ],
        ...
    ],

# Usage
See the [USECASES.md](https://github.com/wdmg/yii2-users/blob/master/USECASES.md) for more details.

# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('users')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.1.2.7 - Fixed identity, add user options
* v.1.2.6 - Fixed copyrights and getUserId()
* v.1.2.5 - Fixed: syntax error, unexpected '::' (T_PAAMAYIM_NEKUDOTAYIM)
* v.1.2.4 - Multiple Sign In`s and logoffs by session timeout

CHANGELOG.md
Changelog
=========

## 1.2.7 (2023-06-20)
 * Fixed identity, add user options

## 1.2.6 (2023-06-18)
 * Fixed getUserId()
 * Fixed copyrights

## 1.2.5 (2021-06-20)
 * Fixed: syntax error, unexpected '::' (T_PAAMAYIM_NEKUDOTAYIM)

## 1.2.4 (2021-05-02)
 * Fixed: check multiple sign in, added counter stats method
 * Multiple Sign In`s and logoffs by session timeout
 
## 1.2.3 (2021-01-01)
 * Unsupported array function fixed
 
## 1.2.2 (2020-12-24)
 * Bugfix when RestAPI request
 
## 1.2.1 (2020-10-07)
 * Bugfix when save user password
 
## 1.2.0 (2020-09-02)
 * User is online feature
 * Filter users by role, assign/revoke user role
 * CRUD for add/update user info
 * Added getUsername() and getEmail() methods
 
## 1.1.8 (2020-03-27)
 * Added pagination
 * Up to date dependencies
 * Fixed SlectInput in GridView
 * Added statuses list in view, translations
 
## 1.1.7 (2019-12-10)
 * Fixed deprecated class declaration

## 1.1.6 (2019-07-16)
 * Added extra options to composer.json and navbar menu icon
 
## 1.1.5 (2019-06-09)
 * Added choice param for non interactive mode
 
## 1.1.4 (2019-06-07)
 * Module refactoring
 
## 1.1.3 (2019-06-05)
 * Module transferred to base module interface. Update Yii2 version.

## 1.1.2 (2019-06-01)
 * Bugfix, routing and refactoring
 
## 1.1.1 (2019-05-31)
 * Update Yii2 version

## 1.1.0 (2019-05-28)
 * Added module name and description, translations. Routing fix.
 
## 1.0.4 (2019-03-16)
 * Added dashboard navigation items for NavBar
 
## 1.0.3 (2019-02-28)
 * Added module priority of initialization
 * Fixing tables names in migrations
 
## 1.0.2 (2019-01-29)
 * Registration confirmation functionality
 * Added module options

## 1.0.1 (2019-01-27)
 * Bugfix and refactoring
 * Added email template for restore password
 * Added username custom validation
 * Autosign user role when signup
 * Signup & signin functionality
 
## 1.0.0 (2019-01-15)
 * First pre-release
 * Added UI translations
 * Complete view for module
 * Update Bootstrap and added URL rules
 * Added base CRUD for module
 * Added base model class
 * Added batch insert demo data

Module.php
<?php

namespace wdmg\users;

/**
 * Yii2 Users
 *
 * @category        Module
 * @version         1.2.7
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-users
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use Yii;
use wdmg\base\BaseModule;
use yii\helpers\ArrayHelper;

/**
 * Users module definition class
 */
class Module extends BaseModule
{

    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\users\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = "users/index";

    /**
     * @var string, the name of module
     */
    public $name = "Users";

    /**
     * @var string, the description of module
     */
    public $description = "Users management module";

    /**
     * @var string the module version
     */
    private $version = "1.2.7";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 2;

    public $rememberDuration = (3600 * 24 * 30);
    public $signupConfirmation = [
        'needConfirmation' => false,
        'checkTokenRoute' => 'site/signup-confirm',
        'supportEmail' => 'noreply@example.com',
        'emailViewPath' => [
            'html' => '@vendor/wdmg/yii2-users/mail/signupConfirmation-html',
            'text' => '@vendor/wdmg/yii2-users/mail/signupConfirmation-text',
        ],
    ];
    public $passwordReset = [
        'resetTokenExpire' => 3600,
        'checkTokenRoute' => 'site/reset-password',
        'supportEmail' => 'noreply@example.com',
        'emailViewPath' => [
            'html' => '@vendor/wdmg/yii2-users/mail/passwordReset-html',
            'text' => '@vendor/wdmg/yii2-users/mail/passwordReset-text',
        ],
    ];

    /**
     * @var bool, of allow to multi Sign In
     */
    public $multiSignIn = true;

    /**
     * @var integer, session timeout in sec. of auth (where `0` is unlimited)
     */
    public $sessionTimeout = 0;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);

    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = false)
    {
        $items = [
            'label' => $this->name,
            'url' => [$this->routePrefix . '/'. $this->id],
            'icon' => 'fa fa-fw fa-user',
            'active' => in_array(\Yii::$app->controller->module->id, [$this->id])
        ];
        return $items;
    }

    /**
     * {@inheritdoc}
     */
    public function bootstrap($app)
    {
        parent::bootstrap($app);

        if (isset(Yii::$app->params['users.multiSignIn']))
            $this->multiSignIn = Yii::$app->params['users.multiSignIn'];

        if (isset(Yii::$app->params['users.sessionTimeout']))
            $this->sessionTimeout = Yii::$app->params['users.sessionTimeout'];

        if (!($app instanceof \yii\console\Application) && !$app->user->isGuest) {
            $module = $this;
            \yii\base\Event::on(\yii\web\Controller::class, \yii\web\Controller::EVENT_BEFORE_ACTION, function ($event) use ($app, $module) {
                if ($app->user->identity) {
                    $lastseen_at = $app->user->identity->lastseen_at;
                    if (strtotime('-1 minutes', strtotime(date('Y-m-d H:i:s'))) > strtotime($lastseen_at)) {
                        $app->user->identity->updateAll(['lastseen_at' => date('Y-m-d H:i:s')], ['id' => $app->user->id]);
                    } else if (intval($module->sessionTimeout) > 0) {
                        $time_delta = intval(strtotime(date('Y-m-d H:i:s')) - strtotime($lastseen_at)) - intval($module->sessionTimeout);

                        if ($time_delta >= intval($module->sessionTimeout)) {

                            $timeout = intval($module->sessionTimeout);
                            $hours = floor($timeout / 3600);
                            $minutes = floor(($timeout / 60) % 60);
                            $seconds = $timeout % 60;

                            if ($hours > 0) {
                                $timeout_msg = Yii::t('app/modules/users', 'Session was terminated automatically due to inactivity for more than {timeout, plural, =0{} =1{# hour} one{# hour} few{# hours} many{# hours} other{# hours}}.', [
                                    'timeout' => $hours
                                ]);
                            } else if ($minutes > 0) {
                                $timeout_msg = Yii::t('app/modules/users', 'Session was terminated automatically due to inactivity for more than {timeout, plural, =0{} =1{# minute} one{# minute} few{# minutes} many{# minutes} other{# minutes}}.', [
                                    'timeout' => $minutes
                                ]);
                            } else {
                                $timeout_msg = Yii::t('app/modules/users', 'Session was terminated automatically due to inactivity for more than {timeout, plural, =0{} =1{# second} one{# second} few{# seconds} many{# seconds} other{# seconds}}.', [
                                    'timeout' => $seconds
                                ]);
                            }

                            $app->user->logout();
                            $app->getSession()->setFlash(
                                'warning',
                                $timeout_msg
                            );
                            $app->controller->goHome();
                        }
                    }
                }
            });
        }
    }
}
